### ISSUE TITLE
Spike: Configurable UI

### ISSUE THREAD

**Issue by AlexGilleran** [magda-io/magda#1615]
At some point we'll have to change to a UI that's customisable. We've got to decide

- How customisable it will be
- How users can customise it
- How this customisation works in the architecture of the frontend

Customisation can be divided into theming and adding extra components, although these overlap slightly.

## Theming
Alternatives:
- Provide semantic HTML and allow for custom CSS to change this in a [CSS Zen Garden](http://www.csszengarden.com/) style approach
    - Allows pretty much infinite flexibility around look and feel
    - How do we make sure that upgrades don't result in broken CSS?
    - How do we make this work with DTA Design Components? They require the markup to be a certain way
    - How do we make it so that plugged-in components work with the same theme?
- Similar to the above, but with _SCSS_
    - Allows compilation which might alleviate the upgrades-breaking-the-css problem a bit (but not completely by any means)
    - Variables mean that there's a clear way to specify themes, which component authors can then lean on
    - Means we'd have to compile into CSS on the fly, which is... complicated
- Don't allow CSS customisation, but allow extensive customisation by theming, kind of the way that [material-ui](https://material-ui.com/) works
    - Makes upgrades much safer
    - Makes it much easier to author components that will fit into a site's overall theme
    - Means that people have to figure out a new interface for theming of our own design which will probably be even worse than CSS
    - We'll never be able to make it customisable enough to please everyone - at some point every magda instance will look magda-ey

## Plugins
Alternatives:
- [Micro-frontends](https://micro-frontends.org/)
    - Fits into our whole language-agnostic-extension philosophy
    - Means we could maybe put the frontend into microservices 💥 
    - Way more complicated than nearly everything else
    - How do you apply a consistent theme to them?
    - How do we communicate between components?
    - How would we do server rendering for them?
- Build our own plugin architecture
    - Have people upload some bundle of javascript and css and whatever else they need and then somehow serve it on the fly
    - How do we specify an interface for this?
    - How do we ensure compatibility across upgrades?
-  Enforce React for all our components
    - Makes interfaces simple to specify
    - Makes server rendering much easier
    - How exactly do we integrate external components with our code? Compile on the fly?

## Good ideas we should do no matter what
- Separate model layer out into something that can be used from any framework
    - Means you could create a whole new frontend and not have to worry about the model
    - How should we do this though? Redux? Something simpler like a pub sub model? Something else?

**Comment by t83714** [comment]
Good summary~ 💯 👍

**Comment by AlexGilleran** [comment]
Spent a few days investingating this.

Even though it's new and weird I think the micro-frontends approach syncs really well for us, given that we've already decided that adding microservices is how we want to support extension.

Trying to get micro services working in the front-end is a bit more interesting, especially if we want to leave the possibility of server-side rendering (important for SEO, which currently isn't great with Magda) alive.

It's definitely _possible_ (see https://github.com/ranjanrajiv00/micro-frontends), the question is how to make it easy for us to work with.

My preference for SSR front-end frameworks is Next JS, but that definitely doesn't support this approach at the moment: https://github.com/zeit/next.js/issues/6040. With a bit of work we could probably patch it in and cross our fingers it gets accepted upstream.

Another alternative is to use a next alternative like https://github.com/jaredpalmer/after.js, which is a bit more configurable and could probably be bent to our needs more easily. However the "Getting Started" guide doesn't build and and issue raised about it weeks ago hasn't been acknowledged by the maintainers, so I'm a bit concerned about the viability of that library going forward.

I also tried a SSR web components framework (skatejs) but once again the examples didn't run, and the readme had the wrong command to build a new app `$ skate` vs `$ skatejs` which didn't make a great impression. 

We could also fork an SSR boilerplate and build our own "framework" (like the first example), but if we could get something like Next to work that'd save us a lot of time.

In conclusion: this is worth doing eventually but it's a bunch of work and I think our current UI is sufficient for what we need to do by EOFY. Hopefully by then we'll know how much investment there's gonna be in the project and we'll know whether it makes sense to make architectural upgrades of this magnitude.

This is pretty fascinating so I'll continue monitoring what happens and looking into it when there's spare time :).

My branch is here: https://github.com/magda-io/magda/tree/spike-configurable-ui
